src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,414 @@
+}
+
+void ClusterDiscovery::registerInZk(zkutil::ZooKeeperPtr & zk, ClusterInfo & info)
+{
+    LOG_DEBUG(log, "Registering current node {} in cluster {}", current_node_name, info.name);
+
+    String node_path = getShardsListPath(info.zk_root) / current_node_name;
+    zk->createAncestors(node_path);
+
+    zk->createOrUpdate(node_path, info.current_node.serialize(), zkutil::CreateMode::Ephemeral);
+    LOG_DEBUG(log, "Current node {} registered in cluster {}", current_node_name, info.name);
+}
+
+void ClusterDiscovery::start()
+{
+    if (clusters_info.empty())
+    {
+        LOG_DEBUG(log, "No defined clusters for discovery");
+        return;
+    }
+    LOG_TRACE(log, "Starting working thread");
+
+    auto zk = context->getZooKeeper();
+    for (auto & [_, info] : clusters_info)
+    {
+        registerInZk(zk, info);
+        if (!updateCluster(info))
+        {
+            LOG_WARNING(log, "Error on updating cluster '{}', will retry", info.name);
+            clusters_to_update->set(info.name);
+        }
+    }
+
+    main_thread = ThreadFromGlobalPool([this]
+    {
+        try
+        {
+            runMainThread();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Caught exception in cluster discovery runMainThread");
+        }
+    });
+}
+
+void ClusterDiscovery::runMainThread()
+{
+    setThreadName("ClusterDiscover");
+    LOG_DEBUG(log, "Worker thread started");
+
+    using namespace std::chrono_literals;
+
+    while (!stop_flag)
+    {
+        /// if some cluster update was ended with error on previous iteration, we will retry after timeout
+        clusters_to_update->wait(5s);
+        for (const auto & [cluster_name, need_update] : clusters_to_update->get())
+        {
+            if (!need_update)
+                continue;
+
+            if (updateCluster(cluster_name))
+            {
+                clusters_to_update->unset(cluster_name);
+                LOG_DEBUG(log, "Cluster '{}' updated sucessfully", cluster_name);
+            }
+            else
+            {
+                LOG_DEBUG(log, "Cluster '{}' are not updated, will retry", cluster_name);
+            }
+        }
+    }
+    LOG_DEBUG(log, "Worker thread stopped");
+}
+
+void ClusterDiscovery::shutdown()
+{
+    LOG_TRACE(log, "Shutting down");
+
+    stop_flag.exchange(true);
+    if (main_thread.joinable())
+        main_thread.join();
+}
+
+ClusterDiscovery::~ClusterDiscovery()
+{
+    ClusterDiscovery::shutdown();
+}
+
+bool ClusterDiscovery::NodeInfo::parse(const String & data, NodeInfo & result)
+{
+    try
+    {
+        Poco::JSON::Parser parser;
+        auto json = parser.parse(data).extract<Poco::JSON::Object::Ptr>();
+
+        result.address = json->getValue<std::string>("address");
+        result.secure = json->optValue<bool>("secure", false);
+        result.shard_id = json->optValue<size_t>("shard_id", 0);
+    }
+    catch (Poco::Exception &)
+    {
+        return false;
```

Comment:
I think ignoring exceptions is a bad approach in general.
At least we should log it.
Better, with clear message which filed of which cluster was not parsed.

Reply:
added exception to log